### PR TITLE
Lower IO buffer sizes and max compressed block size (better cache locality)

### PR DESCRIPTION
**Pull request by alexey-milovidov**
Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Lower IO buffer sizes and max compressed block size for better cache locality. Note that some datasets may compress worse with these settings.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
@azat Failed `custom_tld` test most likely indicated an error in the implementation.

**Comment by azat** [comment]
>@azat Failed custom_tld test most likely indicated an error in the implementation.

Indeed, thanks! - #26948

**Comment by alexey-milovidov** [comment]
Lowered memory usage as a bonus.

**Review by kitaisreal — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
@azat Integration test `test_distributed_directory_monitor_split_batch_on_failure` maybe depends on subtle differences in memory usage.

**Comment by alexey-milovidov** [comment]
0.3% performance increase in average :)

**Comment by alexey-milovidov** [comment]
Up to 5 times performance increase and 2 times decrease on specific queries.

**Comment by azat** [comment]
@alexey-milovidov 

New limits:
- `01730_distributed_group_by_no_merge_order_by_long` - `82Mi` (checked only w/o sanitizers)
- `test_distributed_directory_monitor_split_batch_on_failure` - `25Mi`

**Comment by alexey-milovidov** [comment]
Ok. Now what is remaining:

1. Test the most regressed queries in performance test.
Although performance test is already good in overall.

2. Test on a server with HDD.
With queries just after dropping cache and queries with direct IO.

3. Test on a VM with EBS volume (network disk in Yandex.Cloud).
With queries just after dropping cache and queries with direct IO.

File: tests/integration/test_distributed_directory_monitor_split_batch_on_failure/test.py [hunk 1]
```diff
@@ -41,7 +41,7 @@ def test_distributed_directory_monitor_split_batch_on_failure_OFF(started_cluste
         node2.query(f'insert into dist select number/100, number from system.numbers limit {limit} offset {limit*i}', settings={
             # max_memory_usage is the limit for the batch on the remote node
             # (local query should not be affected since 30MB is enough for 100K rows)
-            'max_memory_usage': '30Mi',
+            'max_memory_usage': '25Mi',
```
Comment (azat) [inline tests/integration/test_distributed_directory_monitor_split_batch_on_failure/test.py:44]: You forget to update `test_distributed_directory_monitor_split_batch_on_failure_ON` too (and a comment above)

**Comment by azat** [comment]
@alexey-milovidov some changes for other tests:
- `01641_memory_tracking_insert_optimize`: `max_memory_usage='6Mi'`
- `01598_memory_limit_zeros`: `max_untracked_memory='500Ki'` (note, that failed allocation is still from the ZerosSource::createColumn, not in some basic buffers for query, or similar)

**Comment by alexey-milovidov** [comment]
The most regressed query is actually faster when I test on machine:

```
clickhouse-benchmark <<< "SELECT RegionID, uniq(UserID) AS u FROM test.hits GROUP BY RegionID ORDER BY u DESC LIMIT 10"

New:

  20,62%  [kernel]                               [k] acpi_processor_ffh_cstate_enter
  16,73%  clickhouse                             [.] UniquesHashSet<DefaultHash<unsigned long, void> >::insert
   8,12%  clickhouse                             [.] DB::Aggregator::executeImplBatch<false, false, DB::AggregationMethodOneNumber<unsigned int, HashMapTable<unsigned long, HashMapCell<unsigned long, char*, HashCRC32<unsigned long>, HashTableNoSt
   4,48%  [kernel]                               [k] acpi_idle_do_entry
   3,67%  clickhouse                             [.] DB::IAggregateFunctionHelper<DB::AggregateFunctionUniq<unsigned long, DB::AggregateFunctionUniqUniquesHashSetData> >::addBatch
   2,53%  clickhouse                             [.] LZ4::(anonymous namespace)::decompressImpl<32ul, false>
   2,20%  clickhouse                             [.] LZ4::(anonymous namespace)::decompressImpl<16ul, true>
   2,06%  clickhouse                             [.] LZ4::(anonymous namespace)::decompressImpl<8ul, true>
   2,05%  clickhouse                             [.] LZ4::(anonymous namespace)::decompressImpl<16ul, false>
   1,97%  clickhouse                             [.] UniquesHashSet<DefaultHash<unsigned long, void> >::resize
   1,72%  clickhouse                             [.] UniquesHashSet<DefaultHash<unsigned long, void> >::merge
   1,15%  [kernel]                               [k] copy_user_generic_string
   0,82%  clickhouse                             [.] memcpy

Queries executed: 4092.

localhost:9000, queries 4092, QPS: 33.366, RPS: 296088778.352, MiB/s: 3388.467, result RPS: 333.663, result MiB/s: 0.004.

0.000%          0.026 sec.
10.000%         0.029 sec.
20.000%         0.029 sec.
30.000%         0.029 sec.
40.000%         0.030 sec.
50.000%         0.030 sec.
60.000%         0.030 sec.
70.000%         0.030 sec.
80.000%         0.031 sec.
90.000%         0.032 sec.
95.000%         0.032 sec.
99.000%         0.034 sec.
99.900%         0.037 sec.
99.990%         0.038 sec.


Old:

  20,41%  [kernel]                           [k] acpi_processor_ffh_cstate_enter
  16,89%  clickhouse                         [.] UniquesHashSet<DefaultHash<unsigned long, void> >::insert
   8,28%  clickhouse                         [.] DB::Aggregator::executeImplBatch<false, false, DB::AggregationMethodOneNumber<unsigned int, HashMapTable<unsigned long, HashMapCell<unsigned long, char*, HashCRC32<unsigned long>, HashTableNoState>
   6,90%  [kernel]                           [k] copy_user_generic_string
   4,94%  [kernel]                           [k] acpi_idle_do_entry
   3,68%  clickhouse                         [.] DB::IAggregateFunctionHelper<DB::AggregateFunctionUniq<unsigned long, DB::AggregateFunctionUniqUniquesHashSetData> >::addBatch
   2,66%  clickhouse                         [.] LZ4::(anonymous namespace)::decompressImpl<32ul, false>
   2,33%  clickhouse                         [.] LZ4::(anonymous namespace)::decompressImpl<16ul, true>
   2,18%  clickhouse                         [.] LZ4::(anonymous namespace)::decompressImpl<16ul, false>
   2,15%  clickhouse                         [.] LZ4::(anonymous namespace)::decompressImpl<8ul, true>
   2,02%  clickhouse                         [.] UniquesHashSet<DefaultHash<unsigned long, void> >::resize
   1,73%  clickhouse                         [.] UniquesHashSet<DefaultHash<unsigned long, void> >::merge
   0,88%  clickhouse                         [.] memcpy
   0,66%  clickhouse                         [.] CityHash_v1_0_2::CityHash128WithSeed

Queries executed: 4237.

localhost:9000, queries 4237, QPS: 32.060, RPS: 284494471.402, MiB/s: 3255.781, result RPS: 320.597, result MiB/s: 0.004.

0.000%          0.027 sec.
10.000%         0.030 sec.
20.000%         0.030 sec.
30.000%         0.030 sec.
40.000%         0.031 sec.
50.000%         0.031 sec.
60.000%         0.031 sec.
70.000%         0.032 sec.
80.000%         0.032 sec.
90.000%         0.033 sec.
95.000%         0.034 sec.
99.000%         0.036 sec.
99.900%         0.039 sec.
99.990%         0.089 sec.
```

The function `copy_user_generic_string` (copying data from page cache to userspace) becomes much more cheap due to cache locality.

**Comment by alexey-milovidov** [comment]
Major performance degradation on a server with slow HDD RAID:

```
Before:

$ for i in {1..10}; do echo 3 | sudo tee /proc/sys/vm/drop_caches && clickhouse-client --time --query "SELECT sum(ignore(*)) FROM test.hits"; done3
0
12.185
3
0
12.616
3
0
12.473
3
0
12.827
3
0
12.854
3
0
11.613
3
0
11.738
3
0
12.385
3
0
13.205
3
0
13.629

After:

$ for i in {1..10}; do echo 3 | sudo tee /proc/sys/vm/drop_caches && clickhouse-client --time --query "SELECT sum(ignore(*)) FROM test.hits"; done
3
0
24.009
3
0
20.183
3
0
21.827
3
0
19.322
3
0
19.643
3
0
20.333
3
0
18.532
3
0
20.282
3
0
17.893
3
0
19.576
```

Not ready for merge.

**Comment by alexey-milovidov** [comment]
We need to adjust buffer size automatically.

Introduce `min_read_buffer_size` setting, set it to 64K by default.

Determine if block device under filesystem is rotational or not and its optimal IO size:

1. Found mount point of the filesystem with `statvfs`.
2. Read `/proc/self/mountinfo` and find a row with this mount point.
3. Extract device number, e.g. `9:2` from `74 21 9:2 / /opt rw,noatime shared:29 - ext4 /dev/md2 rw,nobarrier,stripe=1792,data=ordered`.
4. Read `/sys/dev/block/9:2/queue/optimal_io_size` and `/sys/dev/block/9:2/queue/rotational`.

If `optimal_io_size` is non zero, cap it with `min_read_buffer_size` and `max_read_buffer_size` - this is the buffer size to use.
If it is zero and device is rotational, choose `max_read_buffer_size`, otherwise `min_read_buffer_size`.

@azat Comments?
Maybe there is more simple way to check if we are working with HDD?

**Comment by Algunenano** [comment]
> Extract device number, e.g. 9:2 from 74 21 9:2 / /opt rw,noatime shared:29 - ext4 /dev/md2 rw,nobarrier,stripe=1792,data=ordered.
> Read /sys/dev/block/9:2/queue/optimal_io_size and /sys/dev/block/9:2/queue/rotational.

There is a missing step there which is to go from partition to disk/device if needed. At least in my machine (5.13.8) those values are only available per disk, nor per partition.

File: src/Core/Defines.h [hunk 2]
```diff
@@ -20,7 +20,7 @@
 #define DBMS_DEFAULT_POLL_INTERVAL 10
 
 /// The size of the I/O buffer by default.
-#define DBMS_DEFAULT_BUFFER_SIZE 1048576ULL
+#define DBMS_DEFAULT_BUFFER_SIZE 65536ULL
```
Comment (azat) [inline src/Core/Defines.h:23]: This may be not optimal for s3

**Comment by azat** [comment]
>PS. I've already tested it on a machine with single Optane SSD on queries with 1 and 16 threads and with querying single column and 100 columns, with cached and direct IO - performance is increased about 20%

Interesting, I've tried to check simply with dd (with 33GiB file):

type|direct|block size|speed
-|-|-|-
ssd|true|1M|527MB/s
ssd|true|64K|445MB/s
ssd|false|1M|4.3GB/s
ssd|false|64K|2.1GB/s
ssd|false-drop-cache|1M|1.0GB/s
ssd|false-drop-cache|64K|1.1GB/s

At first I though about page cache filling, but it is not that simple.
After I though about read_ahead, and it is indeed related (see below).

>Major performance degradation on a server with slow HDD RAID:

Hm, I see degradation with this patch even for ssd (first two snippets in details).
But after tunning read_head (`/sys/block/BLK/queue/read_ahead_kb`), it works the same, and it is indeed explainable, part of data will be read in background and so queries work faster.
And as usual, benchmarking is hard, and should be done with extra care. Since real production load with busy servers is very different, and I assume that busy servers does not get lots of benefit from read ahead (at least that amount of benefit that will show 20% performance improvement).

<details>

### xeon patched ssd w/o caches read_ahead=1MB

```
$ for i in {1..10}; do ls -d /work3/azat/ch/dataset/hits_100m_obfuscated_v1/data/default/hits_100m_obfuscated/201307_1_96_4/* | xargs -P100 -i bash -c 'drop_file_cache {}' && ./clickhouse client --time --query "SELECT sum(ignore(*)) FROM hits_100m_obfuscated"; done
0
28.704
0
27.939
0
27.825
0
27.900
0
27.868
0
27.823
0
27.820
0
27.889
0
27.863
0
27.801
```

### xeon upstream ssd w/o caches

```
function drop_file_cache()
{
    dd oflag=nocache conv=notrunc,fdatasync count=0 of=$@ >& /dev/null
} && export -f drop_file_cache
for i in {1..10}; do ls -d ch/dataset/hits_100m_obfuscated_v1/data/default/hits_100m_obfuscated/201307_1_96_4/* | xargs -P100 -i bash -c 'drop_file_cache {}' && ./clickhouse client --time --query "SELECT sum(ignore(*)) FROM hits_100m_obfuscated"; done

0
27.711
0
27.753
0
27.644
0
27.587
0
27.546
0
27.628
0
27.786
0
27.630
0
27.698
0
27.630
```

### xeon patched ssd w/o caches

```
$ for i in {1..10}; do ls -d /work3/azat/ch/dataset/hits_100m_obfuscated_v1/data/default/hits_100m_obfuscated/201307_1_96_4/* | xargs -P100 -i bash -c 'drop_file_cache
{}' && ./clickhouse client --time --query "SELECT sum(ignore(*)) FROM hits_100m_obfuscated"; done
0
34.241
0
33.480
0
33.425
0
33.457
0
33.411
0
33.359
0
33.358
0
33.438
0
33.392
0
33.424
```

### xeon upstream ssd

```
$ for i in {1..10}; do ./clickhouse client --time --query "SELECT sum(ignore(*)) FROM hits_100m_obfuscated"; done
0
12.585
0
12.265
0
12.190
0
12.054
0
11.992
0
12.125
0
12.090
0
12.037
0
11.591
0
11.290
```

### xeon patched ssd

```
$ for i in {1..10}; do ./clickhouse client --time --query "SELECT sum(ignore(*)) FROM hits_100m_obfuscated"; done
0
34.106
0
11.766
0
11.845
0
11.875
0
11.941
0
11.959
0
11.651
0
11.601
0
12.039
0
11.813
```

### i7 patched ssd

```
$ for i in {1..10}; do ./clickhouse client --time --query "SELECT sum(ignore(*)) FROM test.hits"; done
0
1.498
0
1.483
0
1.443
0
1.500
0
1.465
0
1.476
0
1.468
0
1.465
0
1.539
0
1.473
```

### i7 upstream ssd

```
$ for i in {1..10}; do ./clickhouse client --time --query "SELECT sum(ignore(*)) FROM test.hits"; done
0
1.551
0
1.482
0
1.454
0
1.541
0
1.565
0
1.455
0
1.517
0
1.497
0
1.488
0
1.547
```

</details>

>The function copy_user_generic_string (copying data from page cache to userspace) becomes much more cheap due to cache locality.

This is interesting, if this is due to cache locality then the underlying storage (hdd, ssd, tmpfs) and page cache should not matter , right?

>Read /sys/dev/block/9:2/queue/optimal_io_size and /sys/dev/block/9:2/queue/rotational.

There is `/sys/block/BLKDEV/queue`, and there is no point in looking at per-partition value since these values are reported by the storage device, so it will be the same for all partitions on that block device.

But note that `optimal_io_size` usually does not reported by storage devices (I've also checked it on server for hdd and ssd, and locally for nvme)

>Lower IO buffer sizes and max compressed block size

But I don't think that 1MB block size worth it (but it may increase compression ratio a little bit, or not a little on some specific data), so maybe it wroth to make it lower (i.e. 64K or 128K - default read_ahead size).

**Comment by alexey-milovidov** [comment]
Ok. Also worth to note that cache locality differece appears only on queries with multiple threads (due to exhausting shared L3 cache).

**Pull request closed without merging**